src/Netscript/killWorkerScript.ts

```diff
@@ -43,11 +43,13 @@ function stopAndCleanUpWorkerScript(ws: WorkerScript): void {
   ws.delayReject?.(new ScriptDeath(ws));
   ws.env.runningFn = "";
 
-  if (typeof ws.atExit === "function") {
+  for (const key of ws.atExit.keys()) {
     try {
-      const atExit = ws.atExit;
-      ws.atExit = undefined;
-      atExit();
+      const atExit = ws.atExit.get(key);
+      //Calling ns.exit inside ns.atExit can lead to recursion
+      //so the handler must be removed immediatly
+      ws.atExit.set(key, undefined);
+      if (typeof atExit == "function") atExit();
```

Comment:
Setting individual keys to undefined one at a time does not make sense to me. Is there a reason this is being done instead of just clearing the Map once after storing the relevant contents?

e.g.

```ts
const atExits = [...ws.atExit.values()];
ws.atExit.clear();
for (const atExit of atExits) {
  try {
    atExit();
  }
  catch (e: unknown) {
    handleUnknownError(e, ws, "Error running atExit function.\n\n");
  }
}
```

Reply:
I made the change as @d0sboots suggested. However, your version iterates the atExit values twice. That might not be optimal if you can just do it in one go.

Reply:
Even if keeping it to the current one-iteration approach, is there a reason to explicitly set the value to "undefined" instead of just deleting that key? That would also allow making it a Map of functions, instead of a Map of function|undefined.

Iterating only once is good if it can have the correct behavior. In my suggestion code I captured the entire array of atExits before running any of them, so that the atExits are "locked" at the time the script is exited. Iterating the Map directly while running player code in between iterations leaves the door open for the player to modify the atExits while the script is already in the exiting phase.

Here's an example player script:

```js
/** @param {NS} ns */
export async function main(ns) {
  function setupExitHandler(n) {
    if (n > 10) return; // This is just to prevent an infinite loop
    ns.tprint(`Setting up exit handler ${n}`)
    ns.atExit(() => {
      ns.tprint(`Running exit handler ${n}`);
      setupExitHandler(n + 1);
    }, n.toString());
  }
  setupExitHandler(0);
  ns.tprint("Exiting script");
  ns.exit();
}
```

And here is the actual output in your current code:
```
test.js: Setting up exit handler 0
test.js: Exiting script
test.js: Running exit handler 0
test.js: Setting up exit handler 1
test.js: Running exit handler 1
test.js: Setting up exit handler 2
test.js: Running exit handler 2
test.js: Setting up exit handler 3
test.js: Running exit handler 3
test.js: Setting up exit handler 4
test.js: Running exit handler 4
test.js: Setting up exit handler 5
test.js: Running exit handler 5
```

Whereas I would expect the output to just be:
```
test.js: Setting up exit handler 0
test.js: Exiting script
test.js: Running exit handler 0
test.js: Setting up exit handler 1
```

During exit handler 0, the player attempts to set up exit handler 1, but event handler 1 does not get run because it wasn't set up prior to script death.

So on the iterate once vs. iterate twice issue, it depends on whether you think the first behavior is correct (where the player can continue adding more exit handlers to the pile while the script is already dying).

Reply:
Sorry I wasn't paying attention to this earlier. Just wanted to chime in that I agree with Snarling 💯 and thus is what I was trying to get at earlier, but I didn't express it properly and ended up being confusing.